external/llvm-project/mlir/include/mlir/ExecutionEngine/SystemDetect.h

```diff
@@ -0,0 +1,54 @@
+//===- SystemDetect.h - ExecutionEngine System Detect ----------*- C++ -*--===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file provides a system detection interface with default CPU detection
+// for the execution engine.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_EXECUTIONENGINE_SYSTEMDETECT_H_
+#define MLIR_EXECUTIONENGINE_SYSTEMDETECT_H_
+
+#include <string>
+#include <unordered_map>
+
+namespace mlir {
+
+/// SystemDetect finds devices on the current system.
+///
+
+struct SystemDevice {
+  enum class Type : int32_t { ECPU, EGPU, ENPU, EALT };
+
+  Type type;
+  uint32_t count;
+  std::string chip;
+  std::unordered_map<std::string, std::string> properties;
+
+  SystemDevice(Type _type = Type::ECPU, const std::string &_chip = {});
+};
+
+class SystemDetect : public std::unordered_map<std::string, SystemDevice> {
```

Comment:
Why inherit from maps? I'd argue it'd be better to provide an API around the map (esp. if we move to the llvm container, which is, last I checked, broadly preferred over std equivalents in a lot of cases)